// esm-exports/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var Path = require("path");
var _ = require("lodash");
var njct_1 = require("njct");
exports.findFileExtensions = ['', '.ts', '.d.ts', '.js', '.tsx', '.jsx'];
exports.fileExtensions2 = ['.ts', '.js'];
exports.fileExtensions3 = ['.tsx', '.jsx'];
function findFile(name, dirname) {
    if (dirname === void 0) { dirname = '.'; }
    var file = null;
    for (var i = 0; i < exports.findFileExtensions.length; i++) {
        var extFile = name + exports.findFileExtensions[i];
        var testFile = Path.resolve(dirname, extFile);
        try {
            var stat = fs.statSync(testFile);
        }
        catch (e) {
            continue;
        }
        if (stat.isDirectory())
            continue;
        file = testFile;
        break;
    }
    return Promise.resolve(file);
}
exports.findFile = findFile;
function uniqEntryList(entryListCollection) {
    return _.chain(entryListCollection)
        .flatten()
        .uniqBy(function (entry) { return entry.hash(); })
        .value();
}
exports.uniqEntryList = uniqEntryList;
// TODO: Test case, map iterator returns false for all, and  directories and not watched
function fileList(basedir, mapIterator, File) {
    if (mapIterator === void 0) { mapIterator = function (path) { return path; }; }
    if (File === void 0) { File = njct_1.inject('fs', function () { return fs; }); }
    var result = [];
    return new Promise(function (resolve) {
        File.readdir(basedir, function (err, files) {
            // Ignore all errors.
            if (!(files && files.length > 0)) {
                return resolve([]);
            }
            var promises = [];
            files.forEach(function (file) {
                var testPath = Path.join(basedir, file).replace(/\\/g, '/');
                promises[promises.length] = new Promise(function (resolve) {
                    File.stat(testPath, function (err, stat) {
                        if (err) {
                            return resolve();
                        }
                        if (stat.isDirectory()) {
                            fileList(testPath, mapIterator)
                                .then(function (r) { return result = result.concat(r); })
                                .then(function () { return resolve(); })
                                .catch(function () { return resolve(); });
                            return;
                        }
                        testPath = mapIterator(testPath);
                        if (testPath) {
                            result.push(testPath);
                        }
                        resolve();
                    });
                });
            });
            var done = function () { return resolve(result); };
            Promise.all(promises)
                .then(done)
                .catch(done);
        });
    });
}
exports.fileList = fileList;
function findEntry(packageDir, _a) {
    var typings = _a.typings, main = _a.main;
    if (typings) {
        return findFile(typings, packageDir);
    }
    if (!main) {
        main = 'index';
    }
    return findFile(main, packageDir);
}
exports.findEntry = findEntry;
function readFile(filePath, options) {
    return new Promise(function (resolve, reject) {
        fs.readFile(filePath, options, function (err, data) {
            if (err) {
                return reject(err);
            }
            resolve(data);
        });
    });
}
exports.readFile = readFile;
function readJson(filepath, File) {
    if (File === void 0) { File = njct_1.inject('fs', function () { return fs; }); }
    return new Promise(function (resolve, reject) {
        File.readFile(filepath, 'utf8', function (err, data) {
            if (err) {
                return reject(err);
            }
            resolve(JSON.parse(data));
        });
    });
}
exports.readJson = readJson;
